src/main/resources/lib/nextjs-proxy/proxy.es6

```diff
@@ -39,14 +40,87 @@ const errorResponse = function (url, status, message, req, renderSingleComponent
     }
 };
 
+function cookiesArrayToObject(array) {
+    const cookies = {};
+    if (array?.length > 0) {
+        array.forEach(cookie => {
+            const indexEq = cookie.indexOf("=");
+            if (indexEq > 0) {
+                const indexSc = cookie.indexOf(";");
+                cookies[cookie.substring(0, indexEq)] = cookie.substring(indexEq + 1, indexSc > 0 ? indexSc : undefined);
+            }
+        });
+    }
+    return cookies;
+}
+
 // lib-http response is different from the one controller awaits
 function okResponse(libHttpResponse) {
     return {
         body: libHttpResponse.body || libHttpResponse.bodyStream,
         status: libHttpResponse.status,
         contentType: libHttpResponse.contentType,
         headers: libHttpResponse.headers,
+        cookies: cookiesArrayToObject(libHttpResponse.cookies),
+    }
+}
+
+function doRequest(req, renderSingleComponent) {
+    const frontendUrl = req.url;
+    const xpSiteUrl = req.headers['xpBaseUrl'];
+
+    const response = httpClientLib.request(req);
+
+    if (Math.floor(response.status / 100) === 3) {
+        // it is a 3xx redirect
+        // http client does not seem to set set-cookie header
+        // so we do it manually instead of followRedirect: true
+        const redirectReq = Object.create(req);
+        redirectReq.url = getFrontendServerUrl() + response.headers['location'];
+        log.debug(`Following redirect to [${redirectReq.url}]:`);
+
+        const setCookie = response.headers['set-cookie'];
+        if (setCookie?.length > 0) {
+            // execute set-cookie as a well-mannered client
+            const cookies = cookiesArrayToObject(setCookie);
+            redirectReq.headers['cookie'] = Object.keys(cookies).map(name => `${name}=${cookies[name]}`).join("; ");
+        }
+
+        return doRequest(redirectReq, renderSingleComponent);
+    }
+
+
+    const isOk = response.status === 200;
+    const isHtml = response.contentType.indexOf('html') !== -1;
+    const isJs = response.contentType.indexOf('javascript') !== -1;
+    const isCss = response.contentType.indexOf('stylesheet') !== -1;
+
+    //TODO: workaround for XP pattern controller mapping not picked up in edit mode
+    const xpSiteUrlWithoutEditMode = xpSiteUrl.replace(/\/edit\//, '/inline/');
+
+    if (isHtml) {
+        response.body = renderSingleComponent
+                        ? getSingleComponentHtml(response.body)
+                        : response.body;
+
+        response.pageContributions = getPageContributionsWithBaseUrl(response, xpSiteUrlWithoutEditMode);
+
+    }
+
+    if (isHtml || isJs || isCss) {
+        response.body = getBodyWithReplacedUrls(req, response.body, xpSiteUrlWithoutEditMode);
     }
+
+    response.postProcess = isHtml
+
+
+    log.info(`<--- [${response.status}]: ${frontendUrl}
```

Comment:
I don't know if we would want every single component rendered to be in the log?
Could we do this on the debug level?

Reply:
It is logged for every request, not every component, but we can change it to debug indeed